crates/common/src/ens.rs

```diff
@@ -0,0 +1,185 @@
+
+/// ENS name or Ethereum Address.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum NameOrAddress {
+    /// An ENS Name (format does not get checked)
+    Name(String),
+    /// An Ethereum Address
+    Address(Address),
+}
+
+impl NameOrAddress {
+    /// Resolves the name to an Ethereum Address.
+    pub async fn resolve<N: Network, T: Transport + Clone, P: Provider<N, T>>(
+        &self,
+        provider: &P,
+    ) -> Result<Address, EnsResolutionError> {
+        match self {
+            NameOrAddress::Name(name) => resolve_name(name, provider).await,
+            NameOrAddress::Address(addr) => Ok(*addr),
+        }
+    }
+}
+
+impl From<String> for NameOrAddress {
+    fn from(name: String) -> Self {
+        NameOrAddress::Name(name)
+    }
+}
+
+impl From<&String> for NameOrAddress {
+    fn from(name: &String) -> Self {
+        NameOrAddress::Name(name.clone())
+    }
+}
+
+impl From<Address> for NameOrAddress {
+    fn from(addr: Address) -> Self {
+        NameOrAddress::Address(addr)
+    }
+}
+
+impl FromStr for NameOrAddress {
+    type Err = <Address as FromStr>::Err;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(addr) = Address::from_str(s) {
+            Ok(NameOrAddress::Address(addr))
+        } else {
+            Ok(NameOrAddress::Name(s.to_string()))
+        }
+    }
+}
+
+/// Error type for ENS resolution.
+#[derive(Debug, thiserror::Error)]
+pub enum EnsResolutionError {
+    /// Failed to resolve ENS registry.
+    #[error("Failed to get resolver from ENS registry: {0}")]
+    EnsRegistryResolutionFailed(String),
+    /// Failed to resolve ENS name to an address.
+    #[error("Failed to resolve ENS name to an address: {0}")]
+    EnsResolutionFailed(String),
+}
+
+/// Returns the ENS namehash as specified in [EIP-137](https://eips.ethereum.org/EIPS/eip-137)
+pub fn namehash(name: &str) -> B256 {
+    if name.is_empty() {
+        return B256::ZERO
+    }
+
+    // Remove the variation selector U+FE0F
+    let name = name.replace('\u{fe0f}', "");
+
+    // Generate the node starting from the right
+    name.rsplit('.')
+        .fold([0u8; 32], |node, label| *keccak256([node, *keccak256(label.as_bytes())].concat()))
+        .into()
+}
+
+/// Returns the reverse-registrar name of an address.
+pub fn reverse_address(addr: Address) -> String {
+    format!("{addr:?}.{ENS_REVERSE_REGISTRAR_DOMAIN}")[2..].to_string()
+}
+
+pub async fn get_resolver<N: Network, T: Transport + Clone, P: Provider<N, T>>(
+    node: B256,
+    provider: &P,
+) -> Result<EnsResolverInstance<N, T, &P>, EnsResolutionError> {
+    let registry = EnsRegistry::new(ENS_ADDRESS, provider);
+    let address = registry
+        .resolver(node)
+        .call()
+        .await
+        .map_err(|err| EnsResolutionError::EnsRegistryResolutionFailed(err.to_string()))?
+        ._0;
+
+    Ok(EnsResolver::new(address, provider))
+}
+
+pub async fn resolve_name<N: Network, T: Transport + Clone, P: Provider<N, T>>(
```

Comment:
can we use an extension trait for all these ens functions?

then we could move this to alloy in a followup